File: Design-Patterns-Using-Cpp/Singleton/DatabaseSingleton.cpp
#include <iostream>
#include "sqlite_modern_cpp.h"

class DatabaseSingleton {
private:
    static DatabaseSingleton* instance;
    sqlite::database db;

    DatabaseSingleton() : db("db.sqlite3") {
        std::cout << "Database created" << std::endl;
    }

public:
    static DatabaseSingleton* getInstance() {
        if (!instance) {
            instance = new DatabaseSingleton();
        }
        return instance;
    }

    void createTable() {
        db << "CREATE TABLE IF NOT EXISTS students (id INTEGER, name TEXT);";
    }

    void addData(int id, const std::string& name) {
        db << "INSERT INTO students (id, name) VALUES (?, ?);" << id << name;
    }

    void display() {
        auto result = db << "SELECT * FROM students;";
        for (auto& row : result) {
            std::cout << row["id"] << " " << row["name"] << std::endl;
        }
    }
};

DatabaseSingleton* DatabaseSingleton::instance = nullptr;

// Client code
int main() {
    DatabaseSingleton* db1 = DatabaseSingleton::getInstance();
    DatabaseSingleton* db2 = DatabaseSingleton::getInstance();

    std::cout << "Database Objects DB1: " << db1 << std::endl;
    std::cout << "Database Objects DB2: " << db2 << std::endl;

    db1->createTable();
    db1->addData(1, "john");
    db2->addData(2, "smith");

    db1->display();

    delete db1;  // Clean up allocated instance

    return 0;
}

File: Design-Patterns-Using-JavaScript/Singleton/DatabaseSingleton.js
const sqlite3 = require('sqlite3').verbose();

class DatabaseSingleton {
    constructor() {
        console.log('Database created');
        this.connection = new sqlite3.Database('db.sqlite3');
        this.createTable();
    }

    static getInstance() {
        if (!this._instance) {
            this._instance = new DatabaseSingleton();
        }
        return this._instance;
    }

    createTable() {
        this.connection.run(
            'CREATE TABLE IF NOT EXISTS students (id INTEGER, name TEXT);'
        );
    }

    addData(id, name) {
        const statement = this.connection.prepare(
            'INSERT INTO students (id, name) VALUES (?, ?);'
        );
        statement.run(id, name);
    }

    display() {
        this.connection.each('SELECT * FROM students;', (err, row) => {
            if (err) {
                console.error(err.message);
            } else {
                console.log(`${row.id} ${row.name}`);
            }
        });
    }
}

// Client code
const db1 = DatabaseSingleton.getInstance();
const db2 = DatabaseSingleton.getInstance();

console.log('Database Objects DB1:', db1);
console.log('Database Objects DB2:', db2);

db1.addData(1, 'john');
db2.addData(2, 'smith');

db1.display();
